### PR TITLE
[Parser] Add support for parsing prefixed strings

### PR DESCRIPTION
**Pull request by akshanshbhatt**
Related: #512

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
Example Code:
```python
"Hello, {first_name} {last_name}. You are {age}. "
```

**Comment by certik** [comment]
I think that's a good start. We should add tests.

**Comment by certik** [comment]
It seems there is some issue on Apple M1:
```console
$ cat a.py
x = 3
a = f"ab{x}cd"
$ lpython --show-ast a.py
(Module [(Assign [(Name x Store)] (ConstantInt 3 ()) ()) (Assign [(Name a Store)] (JoinedStr [(ConstantStr "ab" ()) (FormattedValue (Name x Load) -1 ()) (ConstantStr "cd" ())]) ())] [])
$ lpython --show-ast --new-parser a.py
(Module [(Assign [(Name x Store)] (ConstantInt 3 ()) ()) (Assign [(Name a Store)] (JoinedStr [(ConstantStr "????Eb" ()) (FormattedValue (Name x Load) -1 ()) (ConstantStr "cd" ())]) ())] [])
```

File: src/lpython/parser/semantics.h [hunk 1]
```diff
@@ -423,16 +423,76 @@ char* concat_string(Allocator &al, ast_t *a, char *b) {
...
 #define INTEGER(x, l) make_ConstantInt_t(p.m_a, l, x, nullptr)
 #define STRING1(x, l) make_ConstantStr_t(p.m_a, l, x.c_str(p.m_a), nullptr)
 #define STRING2(x, y, l) make_ConstantStr_t(p.m_a, l, concat_string(p.m_a, x, y.c_str(p.m_a)), nullptr)
+#define STRING3(id, x, l) PREFIX_STRING(p.m_a, l, name2char(id), x.c_str(p.m_a))
 #define FLOAT(x, l) make_ConstantFloat_t(p.m_a, l, x, nullptr)
 #define COMPLEX(x, l) make_ConstantComplex_t(p.m_a, l, 0, x, nullptr)
 #define BOOL(x, l) make_ConstantBool_t(p.m_a, l, x, nullptr)
 
+static inline ast_t *PREFIX_STRING(Allocator &al, Location &l, char *prefix, char *s){
+    Vec<ast_t *> exprs;
```
Comment (certik) [inline src/lpython/parser/semantics.h:439]:
```suggestion
    Vec<expr_t *> exprs;
```

File: src/lpython/parser/semantics.h [hunk 2]
```diff
@@ -423,16 +423,76 @@ char* concat_string(Allocator &al, ast_t *a, char *b) {
...
+                exprs.push_back(al, tmp);
+            } else {
+                tmp = make_Name_t(al, l,
+                        (char *)strs[i].c_str(), expr_contextType::Load);
+                tmp = make_FormattedValue_t(al, l, EXPR(tmp), -1, nullptr);
+                exprs.push_back(al, tmp);
+            }
+        }
+    }
+    return make_JoinedStr_t(al, l, EXPRS(exprs), exprs.size());
```
Comment (certik) [inline src/lpython/parser/semantics.h:493]:
```suggestion
    return make_JoinedStr_t(al, l, exprs.p, exprs.size());
```

File: src/lpython/parser/semantics.h [hunk 3]
```diff
@@ -423,16 +423,76 @@ char* concat_string(Allocator &al, ast_t *a, char *b) {
...
+                s1.push_back('"');
+                strs.push_back(s1);
+            }
+        }
+
+        for (size_t i = 0; i < strs.size(); i++) {
+            if (strs[i][0] == '"') {
+                strs[i] = strs[i].substr(1, strs[i].length() - 2);
+                tmp = make_ConstantStr_t(al, l, (char *)strs[i].c_str(), nullptr);
+                exprs.push_back(al, tmp);
```
Comment (certik) [inline src/lpython/parser/semantics.h:484]:
```suggestion
                exprs.push_back(al, ASR::down_cast<ASR::expr_t>(tmp));
```

File: src/lpython/parser/semantics.h [hunk 4]
```diff
@@ -423,16 +423,76 @@ char* concat_string(Allocator &al, ast_t *a, char *b) {
...
+            if(i == str.length()-1 && s1 != "\"") {
+                s1.push_back('"');
+                strs.push_back(s1);
+            }
+        }
+
+        for (size_t i = 0; i < strs.size(); i++) {
+            if (strs[i][0] == '"') {
+                strs[i] = strs[i].substr(1, strs[i].length() - 2);
+                tmp = make_ConstantStr_t(al, l, (char *)strs[i].c_str(), nullptr);
```
Comment (certik) [inline src/lpython/parser/semantics.h:483]: You can't do this, as it becomes dangling when `strs` goes out of scope. You have to allocate all strings using our allocator, such as:
```suggestion
                tmp = make_ConstantStr_t(al, l, s2c(al, strs[i]), nullptr);
```

File: src/lpython/parser/semantics.h [hunk 5]
```diff
@@ -423,16 +423,76 @@ char* concat_string(Allocator &al, ast_t *a, char *b) {
...
+                exprs.push_back(al, tmp);
+            } else {
+                tmp = make_Name_t(al, l,
+                        (char *)strs[i].c_str(), expr_contextType::Load);
```
Comment (certik) [inline src/lpython/parser/semantics.h:487]:
```suggestion
                        s2c(al, strs[i]), expr_contextType::Load);
```

**Review by certik — CHANGES_REQUESTED** [review]
The suggestions above fix the dangling pointers issue.

File: src/lpython/parser/semantics.h [hunk 6]
```diff
@@ -423,16 +423,76 @@ char* concat_string(Allocator &al, ast_t *a, char *b) {
     return LFortran::s2c(al, std::string(s) + std::string(b));
 }
 
+
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/parser/semantics.h:426]:
```suggestion
```

**Comment by certik** [comment]
@akshanshbhatt does this work now on M1, or is there still an issue?

**Comment by akshanshbhatt** [comment]
> @akshanshbhatt does this work now on M1, or is there still an issue?

Yeah, the issue is resolved now. Works perfectly fine!

**Review by certik — APPROVED** [review]
Looks great to me, thanks!

@Thirumalai-Shaktivel if you are ok with the PR, you can go ahead and merge it.

**Comment by akshanshbhatt** [comment]
@certik, I just noticed that the expression tree is not generated correctly in the AST. Is there any way we can handle this?

```zsh
❯ cat examples/expr2.py
f'''
a is {a}
b is {b}
sum is {a + b}
'''
❯ lpyast examples/expr2.py
(Module [(Expr (JoinedStr [(ConstantStr "
a is " ()) (FormattedValue (Name a Load) -1 ()) (ConstantStr "
b is " ()) (FormattedValue (Name b Load) -1 ()) (ConstantStr "
sum is " ()) (FormattedValue (BinOp (Name a Load) Add (Name b Load)) -1 ()) (ConstantStr "
" ())]))] [])
❯ lpynewpar examples/expr2.py
(Module [(Expr (JoinedStr [(ConstantStr "
a is " ()) (FormattedValue (Name a Load) -1 ()) (ConstantStr "
b is " ()) (FormattedValue (Name b Load) -1 ()) (ConstantStr "
sum is " ()) (FormattedValue (Name a + b Load) -1 ()) (ConstantStr "
" ())]))] [])
```

**Comment by certik** [comment]
It seems you just have to convert `a+b` into `(BinOp (Name a Load) Add (Name b Load))`, so probably just call `visit_expr()` on it.

**Comment by certik** [comment]
This might be complicated, as we have to run the full parser on it. I recommend for now to ignore this bug, let's get this merged, and we can resolve this later.

**Comment by akshanshbhatt** [comment]
Should I open an issue regarding this?

**Comment by certik** [comment]
Yes, go ahead and open up an issue for this. 

Thirumalai might be busy, but this is ready to be merged as far as I am concerned.

**Comment by akshanshbhatt** [comment]
> Thirumalai might be busy, but this is ready to be merged as far as I am concerned. @akshanshbhatt are you ok if I merge it?

Yeah, sure!

**Comment by certik** [comment]
Thanks!

**Comment by Thirumalai-Shaktivel** [comment]
Thank you @certik for reviewing and merging this PR.
Also, Thank you @akshanshbhatt for making the changes as suggested and adding the tests.

**Pull request merged**